File: csharp/PromptPack.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace PromptPack;

/// <summary>
/// Converts JSON arrays of objects into token-efficient CSV for LLM prompts.
/// </summary>
public static class PromptPack
{
    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /// <summary>
    /// Convert a JSON string to the most token-efficient text format.
    /// Arrays of similar objects become CSV; everything else stays compact JSON.
    /// </summary>
    public static string Pack(string json)
    {
        JsonElement element;
        try
        {
            using var doc = JsonDocument.Parse(json);
            element = doc.RootElement.Clone();
        }
        catch
        {
            return json;
        }

        return Pack(element);
    }

    /// <summary>
    /// Convert a JsonElement to the most token-efficient text format.
    /// </summary>
    public static string Pack(JsonElement element)
    {
        if (element.ValueKind == JsonValueKind.Array && element.GetArrayLength() >= 2)
        {
            var items = new List<JsonElement>();
            foreach (var item in element.EnumerateArray())
                items.Add(item);

            if (IsPackableArray(items))
                return ToCsv(items);
        }

        return ToCompactJson(element);
    }

    /// <summary>
    /// Convert packed text (CSV or JSON) back to a JsonElement.
    /// </summary>
    public static JsonElement Unpack(string text)
    {
        text = text.Trim();
        if (string.IsNullOrEmpty(text))
        {
            using var emptyDoc = JsonDocument.Parse("[]");
            return emptyDoc.RootElement.Clone();
        }

        if (text[0] == '{' || text[0] == '[')
        {
            try
            {
                using var doc = JsonDocument.Parse(text);
                return doc.RootElement.Clone();
            }
            catch { }
        }

        return FromCsv(text);
    }

    /// <summary>
    /// Combine a user message with packed data.
    /// </summary>
    public static string PackForPrompt(string message, string json)
    {
        return message + "\n" + Pack(json);
    }

    // -----------------------------------------------------------------------
    // Shape detection
    // -----------------------------------------------------------------------

    private static bool IsPackableArray(List<JsonElement> items)
    {
        if (items.Count < 2) return false;

        // All items must be objects
        foreach (var item in items)
        {
            if (item.ValueKind != JsonValueKind.Object) return false;
        }

        // Compute superset and intersection of keys
        var supersetKeys = new HashSet<string>();
        HashSet<string>? intersectionKeys = null;

        foreach (var item in items)
        {
            var currentKeys = new HashSet<string>();
            var flat = FlattenObject(item, "");
            foreach (var k in flat.Keys)
            {
                currentKeys.Add(k);
                supersetKeys.Add(k);
            }

            if (intersectionKeys == null)
                intersectionKeys = new HashSet<string>(currentKeys);
            else
                intersectionKeys.IntersectWith(currentKeys);
        }

        if (supersetKeys.Count == 0) return false;
        if (intersectionKeys == null || intersectionKeys.Count == 0) return false;

        // Each row must have >= 30% of superset keys
        double threshold = Math.Max(supersetKeys.Count * 0.3, 1);
        foreach (var item in items)
        {
            var flat = FlattenObject(item, "");
            if (flat.Count < threshold) return false;
        }

        return true;
    }

    // -----------------------------------------------------------------------
    // CSV writing
    // -----------------------------------------------------------------------

    private static Dictionary<string, object?> FlattenObject(JsonElement element, string prefix)
    {
        var result = new Dictionary<string, object?>();

        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in element.EnumerateObject())
            {
                string key = string.IsNullOrEmpty(prefix) ? prop.Name : prefix + "." + prop.Name;
                var nested = FlattenValue(prop.Value, key);
                foreach (var kv in nested)
                    result[kv.Key] = kv.Value;
            }
        }

        return result;
    }

    private static Dictionary<string, object?> FlattenValue(JsonElement value, string key)
    {
        var result = new Dictionary<string, object?>();

        switch (value.ValueKind)
        {
            case JsonValueKind.Object:
                foreach (var prop in value.EnumerateObject())
                {
                    string childKey = key + "." + prop.Name;
                    var nested = FlattenValue(prop.Value, childKey);
                    foreach (var kv in nested)
                        result[kv.Key] = kv.Value;
                }
                break;

            case JsonValueKind.Array:
                if (IsPrimitiveArray(value))
                {
                    var parts = new List<string>();
                    foreach (var item in value.EnumerateArray())
                        parts.Add(PrimToStr(item));
                    result[key] = string.Join("|", parts);
                }
                else
                {
                    result[key] = ToCompactJson(value);
                }
                break;

            case JsonValueKind.Null:
            case JsonValueKind.Undefined:
                result[key] = null;
                break;

            case JsonValueKind.True:
                result[key] = true;
                break;

            case JsonValueKind.False:
                result[key] = false;
                break;

            case JsonValueKind.Number:
                if (value.TryGetInt64(out long l))
                    result[key] = l;
                else
                    result[key] = value.GetDouble();
                break;

            case JsonValueKind.String:
                result[key] = value.GetString();
                break;
        }

        return result;
    }

    private static string ToCsv(List<JsonElement> items)
    {
        var flatRows = new List<Dictionary<string, object?>>();
        foreach (var item in items)
            flatRows.Add(FlattenObject(item, ""));

        // Ordered superset of keys (insertion order)
        var headers = new List<string>();
        var seen = new HashSet<string>();
        foreach (var row in flatRows)
        {
            foreach (var k in row.Keys)
            {
                if (seen.Add(k))
                    headers.Add(k);
            }
        }

        var sb = new StringBuilder();
        sb.Append(CsvLine(headers.Select(h => h).ToArray()));

        foreach (var row in flatRows)
        {
            sb.Append('\n');
            var cells = headers.Select(h =>
                row.TryGetValue(h, out var val) ? FormatCell(val) : "").ToArray();
            sb.Append(CsvLine(cells));
        }

        return sb.ToString();
    }

    private static string FormatCell(object? value)
    {
        if (value == null) return "";
        if (value is bool b) return b ? "true" : "false";
        if (value is long l) return l.ToString(CultureInfo.InvariantCulture);
        if (value is int i) return i.ToString(CultureInfo.InvariantCulture);
        if (value is double d) return FormatDouble(d);
        if (value is string s) return s;
        return value.ToString() ?? "";
    }

    private static string FormatDouble(double d)
    {
        if (d == Math.Truncate(d) && !double.IsInfinity(d) && !double.IsNaN(d))
            return ((long)d).ToString(CultureInfo.InvariantCulture);
        return d.ToString("G", CultureInfo.InvariantCulture);
    }

    private static string CsvLine(string[] fields)
    {
        return string.Join(",", fields.Select(CsvEscape));
    }

    private static string CsvEscape(string value)
    {
        if (value.Contains(',') || value.Contains('"') ||
            value.Contains('\n') || value.Contains('\r'))
        {
            return "\"" + value.Replace("\"", "\"\"") + "\"";
        }
        return value;
    }

    // -----------------------------------------------------------------------
    // CSV parsing (unpack)
    // -----------------------------------------------------------------------

    private static JsonElement FromCsv(string text)
    {
        var lines = ParseCsvLines(text);
        if (lines.Count < 2)
        {
            using var emptyDoc = JsonDocument.Parse("[]");
            return emptyDoc.RootElement.Clone();
        }

        var headers = lines[0];
        var rows = new List<Dictionary<string, object?>>();

        for (int i = 1; i < lines.Count; i++)
        {
            var csvRow = lines[i];
            if (csvRow.All(c => c == "")) continue;

            var flat = new Dictionary<string, object?>();
            for (int j = 0; j < headers.Count; j++)
            {
                string raw = j < csvRow.Count ? csvRow[j] : "";
                flat[headers[j]] = ParseCell(raw);
            }
            rows.Add(Unflatten(flat));
        }

        // Convert to JsonElement
        string json = SerializeRows(rows);
        using var doc = JsonDocument.Parse(json);
        return doc.RootElement.Clone();
    }

    private static object? ParseCell(string raw)
    {
        if (raw == "") return null;
        if (raw == "true") return true;
        if (raw == "false") return false;

        // Try integer
        if (Regex.IsMatch(raw, @"^-?\d+$"))
        {
            if (long.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out long l))
                return l;
        }

        // Try double
        if (Regex.IsMatch(raw, @"^-?\d+\.\d+$"))
        {
            if (double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture, out double d))
                return d;
        }

        // Pipe-separated (not starting with '[')
        if (raw.Contains('|') && !raw.StartsWith("["))
        {
            var parts = raw.Split('|');
            return parts.Select(ParseCell).ToList();
        }

        // Try JSON
        if (raw.StartsWith("[") || raw.StartsWith("{"))
        {
            try
            {
                using var doc = JsonDocument.Parse(raw);
                return doc.RootElement.Clone();
            }
            catch { }
        }

        return raw;
    }

    private static Dictionary<string, object?> Unflatten(Dictionary<string, object?> flat)
    {
        var result = new Dictionary<string, object?>();
        foreach (var kv in flat)
        {
            var parts = kv.Key.Split('.');
            SetNested(result, parts, 0, kv.Value);
        }
        return result;
    }

    private static void SetNested(Dictionary<string, object?> map, string[] keys, int index, object? value)
    {
        if (index == keys.Length - 1)
        {
            map[keys[index]] = value;
            return;
        }

        if (!map.TryGetValue(keys[index], out var existing) || existing is not Dictionary<string, object?> sub)
        {
            sub = new Dictionary<string, object?>();
            map[keys[index]] = sub;
        }

        SetNested(sub, keys, index + 1, value);
    }

    private static List<List<string>> ParseCsvLines(string text)
    {
        var lines = new List<List<string>>();
        var current = new List<string>();
        var field = new StringBuilder();
        bool inQuotes = false;

        for (int i = 0; i < text.Length; i++)
        {
            char ch = text[i];

            if (inQuotes)
            {
                if (ch == '"')
                {
                    if (i + 1 < text.Length && text[i + 1] == '"')
                    {
                        field.Append('"');
                        i++;
                    }
                    else
                    {
                        inQuotes = false;
                    }
                }
                else
                {
                    field.Append(ch);
                }
            }
            else
            {
                if (ch == '"')
                {
                    inQuotes = true;
                }
                else if (ch == ',')
                {
                    current.Add(field.ToString());
                    field.Clear();
                }
                else if (ch == '\n')
                {
                    current.Add(field.ToString());
                    field.Clear();
                    lines.Add(current);
                    current = new List<string>();
                }
                else if (ch == '\r')
                {
                    // skip
                }
                else
                {
                    field.Append(ch);
                }
            }
        }

        current.Add(field.ToString());
        if (current.Any(c => c != ""))
        {
            lines.Add(current);
        }

        return lines;
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private static bool IsPrimitiveArray(JsonElement array)
    {
        foreach (var item in array.EnumerateArray())
        {
            if (item.ValueKind == JsonValueKind.Object || item.ValueKind == JsonValueKind.Array)
                return false;
        }
        return true;
    }

    private static string PrimToStr(JsonElement value)
    {
        return value.ValueKind switch
        {
            JsonValueKind.Null => "",
            JsonValueKind.True => "true",
            JsonValueKind.False => "false",
            JsonValueKind.Number => value.TryGetInt64(out long l)
                ? l.ToString(CultureInfo.InvariantCulture)
                : value.GetDouble().ToString("G", CultureInfo.InvariantCulture),
            JsonValueKind.String => value.GetString() ?? "",
            _ => value.ToString()
        };
    }

    private static string ToCompactJson(JsonElement element)
    {
        return JsonSerializer.Serialize(element, new JsonSerializerOptions
        {
            WriteIndented = false
        });
    }

    /// <summary>
    /// Serialize List of dictionaries (with possible nesting) to JSON string.
    /// </summary>
    private static string SerializeRows(List<Dictionary<string, object?>> rows)
    {
        var sb = new StringBuilder();
        sb.Append('[');
        for (int i = 0; i < rows.Count; i++)
        {
            if (i > 0) sb.Append(',');
            SerializeObject(sb, rows[i]);
        }
        sb.Append(']');
        return sb.ToString();
    }

    private static void SerializeObject(StringBuilder sb, Dictionary<string, object?> obj)
    {
        sb.Append('{');
        bool first = true;
        foreach (var kv in obj)
        {
            if (!first) sb.Append(',');
            first = false;
            sb.Append(JsonSerializer.Serialize(kv.Key));
            sb.Append(':');
            SerializeValue(sb, kv.Value);
        }
        sb.Append('}');
    }

    private static void SerializeValue(StringBuilder sb, object? value)
    {
        switch (value)
        {
            case null:
                sb.Append("null");
                break;
            case bool b:
                sb.Append(b ? "true" : "false");
                break;
            case long l:
                sb.Append(l.ToString(CultureInfo.InvariantCulture));
                break;
            case int i:
                sb.Append(i.ToString(CultureInfo.InvariantCulture));
                break;
            case double d:
                sb.Append(FormatDouble(d));
                break;
            case string s:
                sb.Append(JsonSerializer.Serialize(s));
                break;
            case List<object?> list:
                sb.Append('[');
                for (int idx = 0; idx < list.Count; idx++)
                {
                    if (idx > 0) sb.Append(',');
                    SerializeValue(sb, list[idx]);
                }
                sb.Append(']');
                break;
            case Dictionary<string, object?> dict:
                SerializeObject(sb, dict);
                break;
            case JsonElement el:
                sb.Append(ToCompactJson(el));
                break;
            default:
                sb.Append(JsonSerializer.Serialize(value.ToString()));
                break;
        }
    }
}

File: go/promptpack.go
package promptpack

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// Pack converts JSON-compatible data to a compact CSV representation when possible.
// Accepts []map[string]interface{} or a JSON string. If the data is a packable array
// of similar objects (>=2 rows), it returns CSV with dot-flattened headers.
// Otherwise it returns compact JSON.
func Pack(data interface{}) string {
	var rows []map[string]interface{}

	switch v := data.(type) {
	case string:
		// Try parsing as JSON
		if err := json.Unmarshal([]byte(v), &rows); err != nil {
			// Not an array of objects, return compact JSON of whatever it is
			var arbitrary interface{}
			if err2 := json.Unmarshal([]byte(v), &arbitrary); err2 != nil {
				return v
			}
			b, _ := json.Marshal(arbitrary)
			return string(b)
		}
	case []map[string]interface{}:
		rows = v
	case []interface{}:
		for _, item := range v {
			if m, ok := item.(map[string]interface{}); ok {
				rows = append(rows, m)
			} else {
				// Not all items are maps, fallback to JSON
				b, _ := json.Marshal(data)
				return string(b)
			}
		}
	default:
		b, _ := json.Marshal(data)
		return string(b)
	}

	if !isPackableArray(rows) {
		b, _ := json.Marshal(rows)
		return string(b)
	}

	// Flatten all rows
	flatRows := make([]map[string]string, len(rows))
	for i, row := range rows {
		flatRows[i] = flattenMap(row, "")
	}

	// Collect all headers (superset)
	headerSet := map[string]bool{}
	for _, fr := range flatRows {
		for k := range fr {
			headerSet[k] = true
		}
	}
	headers := make([]string, 0, len(headerSet))
	for k := range headerSet {
		headers = append(headers, k)
	}
	sort.Strings(headers)

	// Write CSV
	var buf bytes.Buffer
	w := csv.NewWriter(&buf)

	w.Write(headers)
	for _, fr := range flatRows {
		record := make([]string, len(headers))
		for j, h := range headers {
			record[j] = fr[h] // empty string if missing
		}
		w.Write(record)
	}
	w.Flush()

	return strings.TrimRight(buf.String(), "\n")
}

// Unpack converts a CSV or JSON string back to []map[string]interface{}.
func Unpack(text string) []map[string]interface{} {
	text = strings.TrimSpace(text)
	if text == "" {
		return nil
	}

	// Try JSON first
	if strings.HasPrefix(text, "[") {
		var result []map[string]interface{}
		if err := json.Unmarshal([]byte(text), &result); err == nil {
			return result
		}
	}

	// Parse as CSV
	r := csv.NewReader(strings.NewReader(text))
	records, err := r.ReadAll()
	if err != nil || len(records) < 2 {
		return nil
	}

	headers := records[0]
	var result []map[string]interface{}

	for i := 1; i < len(records); i++ {
		flat := map[string]interface{}{}
		for j, h := range headers {
			var val string
			if j < len(records[i]) {
				val = records[i][j]
			}
			flat[h] = parseCell(val)
		}
		result = append(result, unflatten(flat))
	}

	return result
}

// PackForPrompt combines a user message with packed data.
func PackForPrompt(message string, data interface{}) string {
	packed := Pack(data)
	return message + "\n\n" + packed
}

// isPackableArray checks if the data is suitable for CSV packing.
// Requires: >=2 rows, all maps, shared keys (intersection > 0),
// each row has >= 30% of the superset keys.
func isPackableArray(rows []map[string]interface{}) bool {
	if len(rows) < 2 {
		return false
	}

	// Compute superset and intersection of keys
	supersetKeys := map[string]bool{}
	var intersectionKeys map[string]bool

	for i, row := range rows {
		if len(row) == 0 {
			return false
		}
		currentKeys := map[string]bool{}
		flat := flattenMap(row, "")
		for k := range flat {
			currentKeys[k] = true
			supersetKeys[k] = true
		}
		if i == 0 {
			intersectionKeys = make(map[string]bool)
			for k := range currentKeys {
				intersectionKeys[k] = true
			}
		} else {
			for k := range intersectionKeys {
				if !currentKeys[k] {
					delete(intersectionKeys, k)
				}
			}
		}
	}

	if len(intersectionKeys) == 0 {
		return false
	}

	supersetCount := len(supersetKeys)
	for _, row := range rows {
		flat := flattenMap(row, "")
		rowKeyCount := len(flat)
		if float64(rowKeyCount)/float64(supersetCount) < 0.3 {
			return false
		}
	}

	return true
}

// flattenMap recursively flattens nested maps using dot notation.
// Primitive arrays are pipe-joined. Nested arrays of objects are JSON-encoded.
func flattenMap(m map[string]interface{}, prefix string) map[string]string {
	result := map[string]string{}
	for k, v := range m {
		key := k
		if prefix != "" {
			key = prefix + "." + k
		}
		switch val := v.(type) {
		case map[string]interface{}:
			for fk, fv := range flattenMap(val, key) {
				result[fk] = fv
			}
		case []interface{}:
			if isPrimitiveArray(val) {
				parts := make([]string, len(val))
				for i, item := range val {
					parts[i] = formatValue(item)
				}
				result[key] = strings.Join(parts, "|")
			} else {
				b, _ := json.Marshal(val)
				result[key] = string(b)
			}
		case nil:
			result[key] = ""
		default:
			result[key] = formatValue(v)
		}
	}
	return result
}

// isPrimitiveArray checks if all elements are non-map, non-array (primitives).
func isPrimitiveArray(arr []interface{}) bool {
	for _, item := range arr {
		switch item.(type) {
		case map[string]interface{}, []interface{}:
			return false
		}
	}
	return true
}

// formatValue converts a value to its string representation.
func formatValue(v interface{}) string {
	switch val := v.(type) {
	case nil:
		return ""
	case bool:
		if val {
			return "true"
		}
		return "false"
	case float64:
		if val == math.Trunc(val) && !math.IsInf(val, 0) && !math.IsNaN(val) {
			return strconv.FormatInt(int64(val), 10)
		}
		return strconv.FormatFloat(val, 'f', -1, 64)
	case int:
		return strconv.Itoa(val)
	case int64:
		return strconv.FormatInt(val, 10)
	case string:
		return val
	default:
		return fmt.Sprintf("%v", v)
	}
}

// parseCell converts a CSV cell string back to a typed Go value.
// empty → nil, "true"/"false" → bool, integers, floats, pipe-separated → []interface{},
// JSON strings → parsed.
func parseCell(s string) interface{} {
	if s == "" {
		return nil
	}
	if s == "true" {
		return true
	}
	if s == "false" {
		return false
	}

	// Try integer
	if i, err := strconv.ParseInt(s, 10, 64); err == nil {
		// Make sure the string exactly matches the integer representation
		// to avoid interpreting floats as ints
		if strconv.FormatInt(i, 10) == s {
			return i
		}
	}

	// Try float
	if f, err := strconv.ParseFloat(s, 64); err == nil {
		if strconv.FormatFloat(f, 'f', -1, 64) == s {
			return f
		}
	}

	// Try JSON (arrays/objects)
	if (strings.HasPrefix(s, "[") && strings.HasSuffix(s, "]")) ||
		(strings.HasPrefix(s, "{") && strings.HasSuffix(s, "}")) {
		var parsed interface{}
		if err := json.Unmarshal([]byte(s), &parsed); err == nil {
			return parsed
		}
	}

	// Try pipe-separated (must have at least one pipe and no commas in the value
	// that would suggest it's just regular text)
	if strings.Contains(s, "|") {
		parts := strings.Split(s, "|")
		result := make([]interface{}, len(parts))
		for i, p := range parts {
			result[i] = parsePrimitive(p)
		}
		return result
	}

	return s
}

// parsePrimitive parses a simple string into a bool, int, float, or string.
func parsePrimitive(s string) interface{} {
	if s == "true" {
		return true
	}
	if s == "false" {
		return false
	}
	if i, err := strconv.ParseInt(s, 10, 64); err == nil {
		if strconv.FormatInt(i, 10) == s {
			return i
		}
	}
	if f, err := strconv.ParseFloat(s, 64); err == nil {
		if strconv.FormatFloat(f, 'f', -1, 64) == s {
			return f
		}
	}
	return s
}

// unflatten converts a flat map with dot-notation keys into a nested map.
func unflatten(flat map[string]interface{}) map[string]interface{} {
	result := map[string]interface{}{}
	for k, v := range flat {
		parts := strings.Split(k, ".")
		setNested(result, parts, v)
	}
	return result
}

// setNested sets a value in a nested map structure given a path of keys.
func setNested(m map[string]interface{}, keys []string, value interface{}) {
	if len(keys) == 1 {
		m[keys[0]] = value
		return
	}
	sub, ok := m[keys[0]]
	if !ok {
		sub = map[string]interface{}{}
		m[keys[0]] = sub
	}
	subMap, ok := sub.(map[string]interface{})
	if !ok {
		subMap = map[string]interface{}{}
		m[keys[0]] = subMap
	}
	setNested(subMap, keys[1:], value)
}
